Add etcd-operator to CI

aanm opened the pull request:
Fixes #5120

Depends on https://github.com/cilium/cilium/pull/5297

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cilium/cilium/5334)
<!-- Reviewable:end -->

aanm committed:
test: add etcd-operator to CI

As etcd-operator integration was released, let's dogfood it as well in
our CI.

Signed-off-by: André Martins <andre@cilium.io>

aanm commented on test/helpers/node.go:
```diff
@@ -149,21 +149,29 @@ type ExecOptions struct {
 
-// Exec returns the results of executing the provided cmd via SSH.
-func (s *SSHMeta) Exec(cmd string, options ...ExecOptions) *CmdRes {
+// ExecWithParms returns the results of executing the provided cmd with the
```
Fixed

eloycoto commented on test/helpers/node.go:
```diff
@@ -149,21 +149,29 @@ type ExecOptions struct {
 	if len(options) > 0 {
 		ops = options[0]
 	}
+	cmdStr := cmd
```
I would not do like this, and I would follow ginkgo standar. IMHO. 

https://github.com/cilium/cilium/blob/0c9728feab4f3354634e8db07512ac4b72a66137/test/helpers/wrappers.go#L52-L54

aanm replied:
@eloycoto that might break if the parameters contain spaces. e.g. `$ cat file with spaces`.

eloycoto replied:
Could you give an example? 

The function will be like: 
`ExecWithParams("cat %s %s %d %q", []string{a,b,c,d})`

I don't see how it can be broken.

raybejjani commented on test/helpers/node.go:
```diff
@@ -159,7 +159,7 @@ func (s *SSHMeta) ExecWithParams(cmd string, cmdParams []string, options ...Exec
 	cmdStr := cmd
 	if len(cmdParams) > 0 {
 		for _, param := range cmdParams {
-			cmdStr += fmt.Sprintf("%q ", param)
```
This line showed up in the `test: add etcd-operator to CI` commit, you probably intended for it to go with the ExecWithParam refactor, maybe?

raybejjani commented on test/helpers/node.go:
```diff
@@ -149,21 +149,29 @@ type ExecOptions struct {
 		ops = options[0]
 	}
+	cmdStr := cmd
+	if len(cmdParams) > 0 {
```
The for should no-op if len == 0, so this if isn't strictly needed

raybejjani reviewed (changes requested):
Just things with the commits

aanm committed:
test: install cfssl in test VMs

Signed-off-by: André Martins <andre@cilium.io>

raybejjani reviewed (dismissed):
Oh, is the k8s: annotate k8s node after IP address allocation commit intended to be in this PR? it's the only commit in https://github.com/cilium/cilium/pull/5428

eloycoto commented on test/provision/helpers.bash:
```diff
@@ -67,3 +67,16 @@ function install_k8s_using_binary {
     systemctl enable kubelet
 }
+
+function install_cfssl {
```
Know that is still WIP, but this should be in the packet-ci-project and the CI build should be updated.

ianvernon replied:
Agreed with Eloy's comment. @aanm please file an issue to track this as well as removal of this code once we update the box version.

eloycoto commented on test/test_suite_test.go:
```diff
@@ -151,6 +151,24 @@ func reportCreateVMFailure(vm string, err error) {
+	const etcdOperatorPath = "../examples/kubernetes/addons/etcd-operator/"
+	deployFile := func(filename string) {
+		cmdRes := kubectl.Apply(helpers.GetFilePath(etcdOperatorPath + filename))
+		cmdRes.ExpectSuccess("Unable to deploy ")
```
"Unable to deploy %s", helpers.GetFilePath(etcdOperatorPath + filename)

aanm committed:
test: use cilium_version/k8s_version path instead k8s_version/cilium_version

As it's unlikely that the same Cilium version patch changes for
different k8s version, it's better to switch the directory format where
the patch files are stored to have fewer files to maintain.

Signed-off-by: André Martins <andre@cilium.io>

eloycoto commented on test/helpers/kubectl.go:
```diff
@@ -1037,6 +1044,37 @@ func (kub *Kubectl) CiliumExecAll(cmd string) error {
+		if err != nil {
+			kub.logger.Infof("Error while getting PodList: %s", err)
+			return false
+		}
```
I know that is still wip, but what happens if the `pl.Items` is 0, it'll end to return true where no pods are in place. I'll add the condition in there. 

`pl` is a bit vague name variable, if you rename to podList will be great ;-)

ianvernon replied:
@eloycoto I don't see a problem with if the `len(pl.Items) == 0`. This means that no Cilium pods are running.

aanm replied:
What Ian said

eloycoto replied:
Then, the function name is not clear. 
`WaitForCiliumInitContainerToFinish` means that the init container should run, if there are no Cilium Pods the init container did not run, and the result should be false, and wait until init container start correctly.

aanm replied:
Fixed

aanm committed:
etcd-operator: update etcd-operator to 3.3.9

Signed-off-by: André Martins <andre@cilium.io>

aanm committed:
pkg/node: propagate local node change to kv-store in a controller

As the connection with etcd might timeout, Cilium should keep retrying,
otherwise the local node will be deleted in the kv-store if
`UpdateLocalKeySync` fails to register the node into the kv-store.

Signed-off-by: André Martins <andre@cilium.io>

eloycoto commented on test/helpers/kubectl.go:
```diff
@@ -1842,6 +1880,78 @@ func (kub *Kubectl) CiliumServicePreFlightCheck() error {
+	deployFile := func(filename string) error {
+		cmdRes := kub.Apply(GetFilePath(etcdOperatorPath + filename))
+		if !cmdRes.WasSuccessful() {
+			return fmt.Errorf("Unable to deploy descriptor of etcd-operator %s: %s", filename, cmdRes.CombineOutput().String())
```
`cmdRes.OutputPrettyPrint()`

eloycoto commented on test/helpers/kubectl.go:
```diff
@@ -1842,6 +1880,78 @@ func (kub *Kubectl) CiliumServicePreFlightCheck() error {
+	kub.Exec(fmt.Sprintf("%s '%s'", GetFilePath(etcdOperatorPath+"tls/certs/gen-cert.sh"), "cluster.local"))
+	kub.Exec(GetFilePath(etcdOperatorPath + "tls/deploy-certs.sh"))
+
+	err := deployFile("00-crd-etcd.yaml")
```
Why not something like this: 

```
	etcdDeploymentFiles := []string{
		"00-crd-etcd.yaml",
		"cilium-etcd-cluster.yaml",
		"cilium-etcd-sa.yaml",
		"cluster-role-binding-template.yaml",
		"cluster-role-template.yaml",
		"deployment.yaml",
	}
	for _, manifest := range etcdDeploymentFiles {
		err := deployFile(manifest)
		if err != nil {
			return err
		}
	}
	return nil
```

where etcdDeploymentFiles should be in the `helpers/cons`

eloycoto commented on test/helpers/kubectl.go:
```diff
@@ -1842,6 +1880,78 @@ func (kub *Kubectl) CiliumServicePreFlightCheck() error {
+	kub.Exec(fmt.Sprintf("%s '%s'", GetFilePath(etcdOperatorPath+"tls/certs/gen-cert.sh"), "cluster.local"))
+	kub.Exec(GetFilePath(etcdOperatorPath + "tls/deploy-certs.sh"))
+
+	err := deleteFile("deployment.yaml")
```
That variables names should be in a const or an array used in the install.

eloycoto commented on test/helpers/kubectl.go:
```diff
@@ -1842,6 +1880,78 @@ func (kub *Kubectl) CiliumServicePreFlightCheck() error {
+	if err != nil {
+		return err
+	}
+	err = deleteFile("cluster-role-template.yaml")
```
On delete, all should be deleted, does not matter if fails or not, and return the error at the end, if not the leftovers config will be in there.

eloycoto commented on test/k8sT/Chaos.go:
```diff
@@ -34,7 +34,13 @@ var _ = Describe("K8sChaosTest", func() {
 		kubectl = helpers.CreateKubectl(helpers.K8s1VMName(), logger)
 
-		err := kubectl.CiliumInstall(helpers.CiliumDefaultDSPatch, helpers.CiliumConfigMapPatch)
+		_ = kubectl.Apply(helpers.DNSDeployment())
```
Why we don't check the status?

aanm replied:
I don't remember but I think it was failing to deploy kube-dns because it was already deployed. I can try to add the error again and check.

aanm replied:
@eloycoto it fails https://jenkins.cilium.io/job/Cilium-PR-Ginkgo-Tests-Validated/6783/testReport/junit/k8s-1/11/K8sChaosTest_Endpoint_can_still_connect_while_Cilium_is_not_running/

eloycoto commented on test/k8sT/Health.go:
```diff
@@ -32,7 +32,13 @@ var _ = Describe("K8sHealthTest", func() {
 		kubectl = helpers.CreateKubectl(helpers.K8s1VMName(), logger)
 
-		err := kubectl.CiliumInstall(helpers.CiliumDefaultDSPatch, helpers.CiliumConfigMapPatch)
+		_ = kubectl.Apply(helpers.DNSDeployment())
```
Why we don't check the status?

ianvernon commented on test/helpers/kubectl.go:
```diff
@@ -1055,6 +1055,37 @@ func (kub *Kubectl) CiliumExecAll(cmd string) error {
+		return true
+	}
+
+	err := WithTimeout(body, "Cilium Init Container was not able to initialize or had a successful run", &TimeoutConfig{Timeout: HelperTimeout})
```
had --> have

ianvernon commented on test/helpers/kubectl.go:
```diff
@@ -1055,6 +1055,37 @@ func (kub *Kubectl) CiliumExecAll(cmd string) error {
+	}
+
+	err := WithTimeout(body, "Cilium Init Container was not able to initialize or had a successful run", &TimeoutConfig{Timeout: HelperTimeout})
+	if err != nil {
```
You can simplify this by doing `return WithTimeout<...>`

ianvernon commented on test/helpers/kubectl.go:
```diff
@@ -1780,6 +1826,39 @@ func (kub *Kubectl) CiliumServicePreFlightCheck() error {
 	return nil
 }
 
+func (kub *Kubectl) DeployETCDOperator() error {
```
Please add documentation.

ianvernon commented on test/helpers/kubectl.go:
```diff
@@ -1826,6 +1826,78 @@ func (kub *Kubectl) CiliumServicePreFlightCheck() error {
+	return deployFile("deployment.yaml")
+}
+
+func (kub *Kubectl) DeleteETCDOperator() error {
```
Same here.

ianvernon commented on test/k8sT/Chaos.go:
```diff
@@ -34,7 +34,13 @@ var _ = Describe("K8sChaosTest", func() {
+		_ = kubectl.Apply(helpers.DNSDeployment())
+
+		// Deploy the etcd operator
+		err := kubectl.DeployETCDOperator()
```
Can you add a `By` here saying that we are deploying etcd operator?

ianvernon commented on test/helpers/kubectl.go:
```diff
@@ -1842,6 +1873,78 @@ func (kub *Kubectl) CiliumServicePreFlightCheck() error {
 }
 
+func (kub *Kubectl) DeployETCDOperator() error {
+	const etcdOperatorPath = "../examples/kubernetes/addons/etcd-operator/"
```
Would be helpful in Deploy / Delete functions to add logs or invocations of `By` to describe what's going on. It will make it easier to debug any potential issues in the future.

aanm replied:
@ianvernon there is no `By` available. The apply already has the logs:

```
// Apply applies the Kubernetes manifest located at path filepath.
func (kub *Kubectl) Apply(filePath string) *CmdRes {
	kub.logger.Debugf("applying %s", filePath)
	return kub.Exec(
		fmt.Sprintf("%s apply -f  %s", KubectlCmd, filePath))
}
```

ianvernon commented on test/k8sT/PoliciesNightly.go:
```diff
@@ -30,8 +30,13 @@ var _ = Describe("NightlyPolicies", func() {
 
-		err := kubectl.CiliumInstall(helpers.CiliumDefaultDSPatch, helpers.CiliumConfigMapPatch)
+		// Deploy the etcd operator
+		err := kubectl.DeployETCDOperator()
```
This can be done as a follow-up, but a lot of `BeforeAll` is becoming quite boilerplate heavy. We should factor this out into a generic test setup function that we can invoke in each `BeforeAll`.

aanm replied:
@ianvernon agreed, should be a "BeforeEachAll" or something like that

eloycoto commented on test/k8sT/Updates.go:
```diff
@@ -108,15 +113,16 @@ func InstallAndValidateCiliumUpgrades(kubectl *helpers.Kubectl, oldVersion, newV
-		ExpectAllPodsTerminated(kubectl)
-		ExpectCiliumReady(kubectl)
+		err = kubectl.WaitForCiliumInitContainerToFinish()
+		ExpectWithOffset(1, err).To(BeNil(), "Cilium %q was not able to be clean up environment", newVersion)
```
This should be moved at the end of the file. This cleanup happens on the AfterEAch and the asserts on AfterEach are not working well. Should happen at the end.

aanm replied:
Can you clarify what you mean?

eloycoto replied:
Move from cleanup function to the original one at the end. The cleanup happens on the AfterEach.

eloycoto reviewed (changes requested):
As I mentioned, I not agree to use things from internet to install. Few comments about Install and Delete ETCD operator, and I'm not sure why dns is not checked when installed.

eloycoto commented on test/helpers/kubectl.go:
```diff
@@ -1842,6 +1876,72 @@ func (kub *Kubectl) CiliumServicePreFlightCheck() error {
+	kub.Exec(fmt.Sprintf("%s '%s'", GetFilePath(etcdOperatorPath+"tls/certs/gen-cert.sh"), "cluster.local"))
+	kub.Exec(GetFilePath(etcdOperatorPath + "tls/deploy-certs.sh"))
+
+	etcdDeploymentFiles := []string{
```
Could you move this files to helpers/cons and use only once? Does not make sense to have the same files two times.

aanm replied:
The order they are installed / deleted is important.

eloycoto replied:
```
for i:=len(etcdDeploymentFiles); i==0, i--{
deleteFile(etcdDeploymentFiles[i])
}
```
???

eloycoto commented on test/helpers/kubectl.go:
```diff
@@ -1842,6 +1879,56 @@ func (kub *Kubectl) CiliumServicePreFlightCheck() error {
+	}
+
+	kub.Exec(fmt.Sprintf("%s '%s'", GetFilePath(etcdOperatorPath+"tls/certs/gen-cert.sh"), "cluster.local"))
+	kub.Exec(GetFilePath(etcdOperatorPath + "tls/deploy-certs.sh"))
```
Why we need to execute this on Delete? I'm not sure

aanm replied:
Bad copy-paste.

eloycoto commented on test/helpers/kubectl.go:
```diff
@@ -1842,6 +1879,56 @@ func (kub *Kubectl) CiliumServicePreFlightCheck() error {
+		return nil
+	}
+
+	kub.Exec(fmt.Sprintf("%s '%s'", GetFilePath(etcdOperatorPath+"tls/certs/gen-cert.sh"), "cluster.local"))
```
Should not these two commands return error if fails?

aanm replied:
Fixed

eloycoto commented on test/k8sT/Updates.go:
```diff
@@ -47,18 +47,23 @@ var _ = Describe("K8sUpdates", func() {
 	AfterAll(func() {
 		_ = kubectl.Apply(helpers.DNSDeployment())
 
+		// Deploy the etcd operator as it was removed by the delete --all
```
Why this is needed on the AfterAll? Asserts on AfterAll are not good for how ginkgo works.

You can remove this one, because the next test will install this, and if fails logs will be collected, and happens on BeforeAll that it's a better place to happens. IMHO.
